### ISSUE TITLE
GO15VENDOREXPERIMENT has issues with transitive dependencies

### ISSUE THREAD

**Issue by ansel1** [tools/godep#324]
Start with an empty go package.  Then try `godep save github.com/tools/godep`.

Note that the `vendor/github.com/tools/godep/Godeps` folder is missing the `_workspace` folder.  If this project were checked out in an empty $GOPATH, it would fail to build, because godep's vendored dependencies are not present.  So mixed a GO15VENDOREXPERIMENT parent project with a dependency that uses Godep and path rewriting doesn't work great.

Fair enough.  That can be worked around by manually copying the _workspace folder in.

The other issue is when the dependency is also using GO15VENDOREXPERIMENT.  In this case, the dependencies full `vendor` tree is vendored along with the dependency.  But if both the parent project and the dependency project both depend on a 3rd dependency, there will be two copies of the 3rd dependency (one under the parent project's vendor folder, and one under the dependency's vendor folder).  At compile time, those will be treated as completely different packages, leading to possible compiler errors, like this:

```
cannot use db.Model(&r) (type *"myapp/vendor/github.com/jinzhu/gorm".DB) as type *"myapp/vendor/mylib/vendor/github.com/jinzhu/gorm".DB in argument to...
```

Dependencies would need to be flattened to resolves this, and it's not clear how to do that with godep.  The workaround for this is manually delete the nested vendor folders after doing `godep save`

**Comment by derekdowling** [comment]
I'm running into the second issue as well. Not sure how to get around that.

**Comment by ansel1** [comment]
to workaround missing _workspace folders, I just manually copy them into my vendor folder.  I only need to do this when I add or update a dependency, and the only one which currently has a nested _workspace is godep itself.

To workaround the nested vendor folders, I wrapped godep save and update in make targets, which do this afterwards:

    find vendor/* -type d -name vendor -exec rm -rf '{}' \;

Here's what my "freeze" make target looks like:

    clean.nested.vendor = find vendor/* -type d -name vendor -exec rm -rf '{}' \; &> /dev/null

    # Vendors new dependencies
    vendor.freeze:
    	GO15VENDOREXPERIMENT=1 godep save $(VENDOR_PACKAGES)
    # nested vendor folders don't work
    	-@$(clean.nested.vendor)

**Comment by derekdowling** [comment]
Yeah, my issue is with `"But if both the parent project and the dependency project both depend on a 3rd dependency"`.

**Comment by ansel1** [comment]
Oh, does the 3rd party dep use import rewriting?  then your screwed. :(

**Comment by freeformz** [comment]
FWIW: Godep should support this as it knowns about both styles, old Godeps with and without rewrites and vendor. I hope to fix that issue "soonish", but don't know when I'll get to it.

**Comment by derekdowling** [comment]
@ansel1 luckily I own both repos so I am working around it by not publishing the libraries vendor folders to avoid this.

**Comment by thiderman** [comment]
Was any progress ever made on this? I just started to use godep, and I immediately ran into this. As it stands right now, this hinders me from using it at all. :(

**Comment by freeformz** [comment]
Closing issues before archiving. Thank You\!
